buildtest/buildsystem/base.py

```diff
@@ -640,3 +644,274 @@ def _build_testcontent(self):
+        self.executable = "%s.exe" % os.path.basename(self.source)
+
+        self.compiler_name = None
+        self.lang = self.detect_lang(self.source)
+
+        # Get GNU Compiler if 'gnu' defined in compiler properties.
+        if self.compiler_recipe.get("gnu"):
+            self.compiler_name = "gnu"
+            gnu = GNUCompiler()
+            # get_compilers() will return a list that returns C, C++, and Fortran GNU wrapper
+            self.cc, self.cxx, self.fc = gnu.get_compilers()
+
+        # Get Intel Compiler
+        elif self.compiler_recipe.get("intel"):
+            self.compiler_name = "intel"
+            intel = IntelCompiler()
+            self.cc, self.cxx, self.fc = intel.get_compilers()
+
+        # Get PGI Compiler
+        elif self.compiler_recipe.get("pgi"):
+            self.compiler_name = "pgi"
+            pgi = PGICompiler()
+            self.cc, self.cxx, self.fc = pgi.get_compilers()
+
+        # Get Cray Compiler
+        elif self.compiler_recipe.get("cray"):
+            self.compiler_name = "cray"
+            cray = CrayCompiler()
+            self.cc, self.cxx, self.fc = cray.get_compilers()
+
+        # Set ldflags, cppflags, cflags, cxxflags, fflags if defined in Buildspec
+        self.ldflags = self.compiler_recipe[self.compiler_name].get("ldflags")
+        self.cppflags = self.compiler_recipe[self.compiler_name].get("cppflags")
+        self.cflags = self.compiler_recipe[self.compiler_name].get("cflags")
+        self.fflags = self.compiler_recipe[self.compiler_name].get("fflags")
+        self.cxxflags = self.compiler_recipe[self.compiler_name].get("cxxflags")
+
+        cmd = self.generate_compile_cmd()
+
+        # every test starts with shebang line
+        lines = [self.shebang]
+
+        # if 'module' defined in Buildspec add modules to test
+        if self.recipe.get("module"):
+            for module in self.recipe.get("module"):
+                lines.append(module)
+
+        # add compile command to test
+        lines.append(" ".join(cmd))
+
+        # build run command that launches generated executable. Format: <exec> <args>
+        # if 'exec_args' is not defined it will return 'None' which means no arguments to be passed when running executable
+        run = [
+            self.executable,
+            self.compiler_recipe[self.compiler_name].get("exec_args"),
+        ]
+
+        # remove 'None' in case 'exec_args' is not defined in Buildspec
+        run = list(filter(None, run))
+
+        # add run line to test
+        lines.append(" ".join(run))
+        return lines
+
+    def detect_lang(self, sourcefile):
+        """This method will return the Programming Language based by looking up  file extension of source file."""
+
+        ext = os.path.splitext(sourcefile)[1]
+
+        # if ext not in self.lang_ext_table then raise an error. This table consist of all file extensions that map to a Programming Language
```

Comment:
You shouldn't need this complicated logic if the compilers are easily separated - the compiler will know a set of languages, and exit if there is mismatch.